### DISCUSSION TITLE
ESP 32 S3 Support with and without I2S

### DISCUSSION THREAD

**Discussion by troth530** [grblHAL/ESP32#107] [General]
Hi, I've been enjoying ESP32, but am interested in seeing if it can get faster with ESP32 S3 such as this.
https://www.amazon.com/LuatOS-Development-Bluetooth-Interfaces-Compatible/dp/B0CDX87YX4

I believe you may have already started the work, do you have an example board file?
I am planning on using SD card which will occupy IO14,16,17,18.  3 axes with PWM for spindle, spindle enable, mist, and limit switches for all 3 axes, and PDN_UART for TMC2209.

> Build 20240115Core:   
ESP32: switched to low-level I/O register access for speed, added WS2812 RGB HAL, completed 32 (or 31?) bit I2S shift register output for ESP32-S3, 16 bit not yet ready.

**Comment by terjeio** [comment]
If you are going to use native USB for the main connection it will be a somewhat faster. Other than that it will be comparable to a plain ESP32.
Native USB and more pins available are the main plusses, a negative is that RMT stepping is only available for up to four steppers vs. eight for the plain ESP32. And I2S stepping was hard to crack but should be ok now, except for 16 bit.

> do you have an example board file?

Only this for now:
[generic_i2s_s3_map.zip](https://github.com/grblHAL/ESP32/files/15047546/generic_i2s_s3_map.zip)

Note that when compiling for the S3 you will have to replace the _sdkconfig_ file with the _sdkconfig.s3 file_. It is possible to specify which sdconfig file to use but I have not figured out how yet. And use the 4.4 framework, I use 4.4.3 but 4.4.6 or later may be ok too.

**Reply by luc-github** [reply]
@terjeio I do not know if it can help you but here is an example of what I do to select specific sdkconfig :
* My main CMakeLists.txt: https://github.com/luc-github/ESP3D-TFT/blob/main/CMakeLists.txt
* One of the sdkconfig selection : https://github.com/luc-github/ESP3D-TFT/blob/main/cmake/targets/ESP32S3_FREENOVE_1_1.cmake#L3

**Comment by troth530** [comment]
@terjeio and @luc-github thank you.

> Native USB and more pins available are the main plusses, a negative is that RMT stepping is only available for up to four steppers vs. eight for the plain ESP32. And I2S stepping was hard to crack but should be ok now, except for 16 bit

.Frankly 4 steppers is all plain ESP32 can allow due to pin limits on RMT with some sacrifices of other features.

Based on the cpu_map file you posted I suspect this is a better candidate
https://www.amazon.com/gp/product/B09CPNRDND

@luc-github does ESP32-TFT work with this board and with grblHAL?
If so how do you connect (USB, UART or I2C).  WebUI is awesome, but I'm interested in full offline as well.
https://www.amazon.com/Microcontroller-Development-Capacitive-Peripherals-Interfaces/dp/B0CP3DYJ5K

Thank you!

**Reply by luc-github** [reply]
@troth530  grbl/grblHAL is not yet supported by ESP3d-TFT, but it is planned, if using ESP32-S3, USB and Serial are available. Currently Only Marlin is supported and streaming is only based on single ping / pong ack 
For grbl/grblHAL, TFT UI is not yet defined (I am starting now), and streaming using buffer instead of single ping pong need to be implemented (mandatory for laser I think)

 if any question on ESP3D-TFT please go to https://github.com/luc-github/ESP3D-TFT or join ESP3D discord, better not do poluate this thread

**Reply by terjeio** [reply]
> Based on the cpu_map file you posted I suspect this is a better candidate

That is the one I use for development.

**Comment by troth530** [comment]
@terjeio grbHAL is awesome and I really like the S3.  I have had lots of issues with clone CP2102 devices on ESP32 devkits. In fact most are clone Silabs devices. 
 
I finally got hardware connected and am seeing some issues.  X motor receives the step signal but not X direction signal, it will move but strangely faster than expected, but will not reverse X direction.  Y doesn't get step signal and I believe there is some fundamental mismatch between pin names for S3, so I haven't tested further.   For build 20240420 attached are my cpu_map (customized bdringv4) and my_machine.h  
[ESP32_S3-troth530.zip](https://github.com/user-attachments/files/17479882/ESP32_S3-troth530.zip)

I had to add my board to trinamic_if.c and 2209 UART does work and M122 returns valid information from the drivers, also SD card does work.  Also maybe gets resolved with other issues but changing $14 does not work as it seems to reset prior to writing to EPROM, so I had to make changes directly in config.h.

For build 20241019, I used the same CPU_Map, my_machine.h, but had to comment out 
#ifdef CONFIG_IDF_TARGET_ESP32S3  in usb_serial.c to compile.  It will compile but will not boot.  

Have you tested the S3 with hardware?  We really appreciate your hard work!
I can move this to an issue if you'd prefer.

**Reply by terjeio** [reply]
>  X motor receives the step signal but not X direction signal

This is because GPIO_NUM_5 is used for both Y step and X direction...

>  Also maybe gets resolved with other issues but changing $14 does not work as it seems to reset prior to writing to EPROM, so I had to make changes directly in config.h.

This will have to wait until I am back home, I only have a dev board at hand and no EEPROM to check with. Saving to flash works so maybe it is a wiring issue at your end? Pins are not I2C capable?
I do have a board at home that has an EEPROM mounted, these are the definitions for that:
```
// Define I2C port/pins
#define I2C_PORT        	I2C_NUM_1
#define I2C_SDA          	GPIO_NUM_3
#define I2C_SCL          	GPIO_NUM_4
#define I2C_CLOCK         	100000
```

> For build 20241019, I used the same CPU_Map, my_machine.h, but had to comment out
#ifdef CONFIG_IDF_TARGET_ESP32S3 in usb_serial.c to compile. It will compile but will not boot.

I'll fix that in the next commit.

> I believe there is some fundamental mismatch between pin names for S3

The pin numbers for the plain ESP32 and the S3 variant may not match at all, after all it is a different processor. So map files has to be tailored for the S3.

**Reply by troth530** [reply]
> This is because GPIO_NUM_5 is used for both Y step and X direction...
> 
@terjeio.  Thank you!  
That's embarrassing, I had Y step wired to GPIO_NUM_6 and neglected to change it in the map file.  All pins seem to work except Step Z and I'm debating whether GPIO 15 and 16 pins are actually available as they are multiplexed with U0RTS and U0CTS (UART0 flow control pins).  

Also the motors are very whiny and make high pitched noises when enabled, but it may be some of the trinamic default settings you have hard coded.  

So at least for the April code things seem to be almost working.  The October code just constantly power cycles once programmed and does not boot.

Thank you again.

**Reply by terjeio** [reply]
> Also the motors are very whiny and make high pitched noises when enabled, but it may be some of the trinamic default settings you have hard coded.

Change [this line](https://github.com/grblHAL/Plugins_motor/blob/985d8bf42a4c548b220b731f5ff4488b9d5d23b5/trinamic.h#L83) to 0 (coolstep)

**Comment by troth530** [comment]
> Change [this line](https://github.com/grblHAL/Plugins_motor/blob/985d8bf42a4c548b220b731f5ff4488b9d5d23b5/trinamic.h#L83) to 0 (coolstep)

Hi @terjeio, thank you.  I'm still looking into this.  In the past I've used Hybrid mode and set TMC2209_TPWM_THRS (TPWMTHRS) per the calculator.  When TSTEP is > TPWMTHRS it goes to spreadcycle.  Also I set TMC2209_COOLSTEP_THRS (TCOOLTHRS) as this controls the lower speed for coolstep and when stallguard diag works even though I don't use sensorless, but the datasheet says coolstep at low speed is not accurate, so I believe a non-zero setting is preferred and higher than TPWMTHRS.  So I use 100 for TCOOLTHRS.  Another thing is that you added PWM_THRESHOLD_VELOCITY and I suspect it's intended to take place of TPWMTHRS (with units of speed in mm/min instead of step time)?  

[TMC2209_Calculations.xlsx](https://github.com/user-attachments/files/17501233/TMC2209_Calculations.xlsx)

Thank you!

**Comment by troth530** [comment]
Hi @terjeio, when setting PWM_THRESHOLD_VELOCITY you need to add member "float steps_per_mm" to motor_settings_t [here ](https://github.com/grblHAL/Plugins_motor/blob/985d8bf42a4c548b220b731f5ff4488b9d5d23b5/trinamic.h#L324).

**Reply by terjeio** [reply]
Why?

**Reply by troth530** [reply]
> Why?

Hi @terjeio If you mean why set PWM_THRESHOLD_VELOCITY...
Setting PWM_THRESHOLD_VELOCITY sets the speed transition from stealth to spread, if PWM_THRESHOLD_VELOCITY is non-zero, it is ignored and will remain in stealth mode for all speeds.   If PWM_THRESHOLD_VELOCITY is set then this [code](https://github.com/grblHAL/Plugins_motor/blob/985d8bf42a4c548b220b731f5ff4488b9d5d23b5/trinamic.c#L363) is active and compiling will fail because struct for motor_settings_t is incomplete.

M913 does not work at run time.  
Thank you!

**Reply by terjeio** [reply]
[Fix committed](https://github.com/grblHAL/Plugins_motor/commit/0cea0494415c8b71b3cf50f7bfc4716866433844) for the invalid settings reference.
Note that I am currently not able to verify this or M913 not working since I do not have access to a Trinamic equipped board until I am back at base.

**Reply by troth530** [reply]
> [Fix committed](https://github.com/grblHAL/Plugins_motor/commit/0cea0494415c8b71b3cf50f7bfc4716866433844) for the invalid settings reference. Note that I am currently not able to verify this or M913 not working since I do not have access to a Trinamic equipped board until I am back at base.

Hi @terjeio 
M913 works, I may have used bad syntax.  Settings seem to check out fine after made the same change you made [Your Fix](https://github.com/grblHAL/Plugins_motor/commit/0cea0494415c8b71b3cf50f7bfc4716866433844)

I may have mis-stated using about control signals and EEPROM on VER:1.1f.20240420.  Actually I'm using this [devKit](https://www.amazon.com/gp/product/B09CPNRDND) and no additional memory.  Maybe I should have called it nvs. 

With this devKit, the control signals need to be hard coded due to crashing when attempting to set $14 (and/or $17) .  
I still cannot test with your latest builds as October build does not even boot.
Thank you and enjoy your time away from the base!

**Reply by terjeio** [reply]
> I still cannot test with your latest builds as October build does not even boot.

I got the same problem suddenly, I had to erase the flash before reflashing to fix it.

**Reply by troth530** [reply]
Hi @terjeio, actually I use a script and erase flash before I program always.  

I found that on the April code (VER:1.1f.20240420) I made  a minor change probably due to compilation issue ... I got it working on S3 after I commented out the section in driver.c near line 2250 in the screenshot below.  This code seems to have been updated since.  After commenting out the same section in the Oct code (VER:1.1f.20241019) I was able to get the S3 to boot, but trinamic did not work.  
![image](https://github.com/user-attachments/assets/f28afe11-f5ab-4d67-abe3-c2e509c83779)

So I will try again with the most recent code.  
BTW, I previously had issues with slow SD card on the ESP32 S1, the SD card writes are super fast on the ESP32 S3 with the same hardware.  But I still seem to have a parsing error grblHAL on WebUI.  But I'll check on the latest code to see if the parsing continues.

Also is this a transcription [error](https://github.com/grblHAL/Plugins_motor/blob/0cea0494415c8b71b3cf50f7bfc4716866433844/trinamic.c#L360C1-L360C55)?  

Thank you!

**Comment by jper1679** [comment]
Hey guy!
I'm quite new to GRBL-HAL, but been using a blackbox since it came out and recently acquired a pretty big router (6x8) which came with Mach3 and my operator asked me if possible to have the same sender... Long story short, I'm in the process of designing a breackout board and was looking at the ESP32-S3 ETH from Waveshare : https://www.waveshare.com/esp32-s3-eth.htm
The rational is that wifi does't work well in my shop and I'd prefer to use the wire ethernet link.
Do you guys know or have tested this board? Would it be possible to use the ethernet port to send the gcode from a computer?
Thanks in advance!

**Reply by troth530** [reply]
Get Phil's Teensy board for Ethernet. Currently trinamic uart is not working with esp32 s3 anyway.

**Comment by troth530** [comment]
Hi @terjeio, so I'm still trying to get all of the S3 working.  I figured out you changed the serial port for trinamic UART.  I am assuming HAS_BOARD_INIT no longer needs to be defined.?

My biggest issue is that now I'm having issues with choosing the correct way to select the spindle.  My application is laser.  I am using PWM with Spindle Enable for laser.  I don't have enough pins for a spindle dir signal and it's not needed for laser.   In my_machine.h I have tried this line
#define SPINDLE0_ENABLE           SPINDLE_PWM0_NODIR
Also, in grbl/config.h I tried this setting for $395, but $395 is not a valid $ setting 
#define DEFAULT_SPINDLE SPINDLE_PWM0_NODIR 
The PWM pin seems to work, but not the Spindle Enable pin. 
$pins reports neither PWM or Spindle enable pin although the PWM pin works.

Also $pins does not report Mist pin (probably flood) pin has the same issue.  

Thank you!

**Reply by terjeio** [reply]
>  I am assuming HAS_BOARD_INIT no longer needs to be defined.?

Yes, not needed if the code only inits the Trinamic UART.

> $pins reports neither PWM or Spindle enable pin although the PWM pin works.
Also $pins does not report Mist pin (probably flood) pin has the same issue.

Which board map? If your own then how pin definitions for spindle and coolant is made has been changed recently.

**Reply by troth530** [reply]
@terjeio Thank you for the hint.  I found the bread crumbs by referring to the latest board maps.  I'll keep testing.  S3 is necessary because most S1 devkits have clone SiLabs chips and they have communications errors.  Lasers are all about speed.

Amazing work you're doing, I'm just trying to keep up.  Cheers!

**Comment by troth530** [comment]
Hi @terjeio it seems with the latest build M913X2000 instead of setting PWM threshold, will set driver current to 2000 temporarily, then M122i sets it back to whatever was set by $140.  I observe this behavior only with GRBL_BUILD 20250114.  The correct behavior for M913X2000 adjusts PWM threshold for X Axis with GRBL_BUILD 20240420.  Same hardware in both cases.

M913X**2000**
ok
M122
[TRINAMIC]
                      X       Y
Driver          TMC2209 TMC2209
Set current        **2000**    1000
RMS current        1768     995
Peak current       2500    1407
...

Attempting to test sensorless homing, but found this issue first.
Thank you!

**Reply by terjeio** [reply]
Oops, a copy/paste error from an earlier refactor. [Fix committed](https://github.com/grblHAL/Plugins_motor/blob/535c6a38fe9900ab35b7db3b369c47bfc97f914d/trinamic.c#L1737-L1743).

**Comment by troth530** [comment]
Hi @terjeio, S3 seems to be fully functioning.  
I found what I believe to be an unrelated [issue](https://github.com/grblHAL/Plugins_motor/issues/20) with sensorless.
Thank you!
